### DISCUSSION TITLE
Is there a way to close certain windows when they are out of focus?

### DISCUSSION THREAD

**Discussion by Prayag2** [hyprwm/Hyprland#3323] [Q&A]
I'm using rofi (xwayland) on my hyprland setup as most plugins that work on rofi don't actually work that well on wofi and other wayland alternatives (I have also tried rofi-wayland but it has major bugs right now). I want it to close when it's out of focus. For example, see this video-

https://github.com/hyprwm/Hyprland/assets/39525869/79a39631-adef-48d7-be7b-73e4a55de09d

In this video, you can see that upon clicking elsewhere, rofi doesn't close. I want it to close when it's out of focus. I also want the same for a few other apps.

**Comment by imxyy1soope1** [comment]
try a shell script mentioned in the [wiki](https://wiki.hyprland.org/IPC/)

**Reply by ghost** [reply]
hyprctl dispatch closewindow (WindowCLASS)

bind = $maindMod, KEY, closewindow, (WindowCLASS)

**Comment by TeChn4K** [comment]
I'm also interested by this feature, but for Wofi. Did you find a solution?

For now, it can't be done by Wofi itself : https://todo.sr.ht/~scoopta/wofi/198

By the way @Prayag2, what tool do you use to create your far right menu on your screenrec?

**Reply by TeChn4K** [reply]
Oh, is it a tray icon with a right click menu ?

**Reply by Prayag2** [reply]
Yes, it's just a tray icon of OBS in waybar :)

**Comment by Gremious** [comment]
This is the top results when searching how to close rofi when unfocused, especially so for wayland, so I thought I'd share my solution.

I'm on plasma wayland, using [lbonn/rofi](https://github.com/lbonn/rofi). The params like `click-to-exit` did not work, so I ended up doing this with a relatively simple script instead.

tl;dr: check active window with `kdotool` and kill rofi if it's not "rofi";

This is for nushell, but it should be easily convertible to any other.

```nu
def rofi [] {
	let ok_str = "OK-for-nushell"
	mut seen_rofi = false;
	mut timeout = 0;
	let cmd = $"rofi -show run -on-entry-accepted \"echo \'($ok_str)\'\"";

	# nushell cannot queue task in the background, so we use `pueue` for that
	# you can use & in bash or what not
	let id = pueue add --immediate $cmd | parse 'New task added (id {id}).'| get id.0;

	loop {
		let curr = kdotool getactivewindow | kdotool getwindowclassname $

		# need this because until rofi actually launches,
		# our current window will be whatever we were on before for 
		if $curr == "rofi" { $seen_rofi = true; }

		if $seen_rofi and $curr != "rofi" {
			let no_selection = pueue log --json $id
				| from json
				| get $"($id)"
				| get output
				| find $ok_str
				| is-empty

			if $no_selection {
				pueue kill $id;
			}

			break;
		}

		sleep 25ms
		$timeout += 25;
		if $seen_rofi == false and $timeout > 250 { break; }
	}
}
```

**Comment by ilivetruth** [comment]
I made these scripts to solve this issue after finding this page and not being able to get it to work. This script works perfectly on hyprland. 

https://github.com/ilivetruth/rofi-wayland-click-to-close
